### DISCUSSION TITLE
Feature Request: Pass 'resources' value to dependsOn PATCH calls for Actions

### DISCUSSION THREAD

**Discussion by puzzledmonkey** [laravel/nova-issues#5549] [Q&A] [answered]
It would be really handy to be able to retrieve the selected resource id/s (or 'all') when constructing the fields($request) array when using dependsOn functionality.

this would fix the final piece of the puzzle when it comes to retrieving the select resource/s ids (or 'all') in the Action fields($request) method. if we have this information, it's possible to generate bespoke fields for a single resource (for example with values pre-filled) or to show different sets of fields when multiple (or all) resources are selected. this is very handy and allows for highly improved UX in some situations.

1. for the initial request in the resource list, it's just a matter of passing in the model as a constructor parameter and storing it in a class property:
```
public function actions(NovaRequest $request)
{
    return [
        new Actions\ActionForResource($this->resource),
        // ...
    ];
}
```

```
class ActionForResource extends Action {
	protected $_model;

	public function __construct($model = null)
	{
		if ($model instanceof Model && $model->id) {
			$this->_model = $model;
		}
	}
```

this allows us to lookup the model for the per-model actions generated in the the initial nova-api/resource/search call.

2. for bulk-select actions, where the user selects one or more checkboxes, or the 'select all' checkbox, we can retrieve the resource ids from $request directly - for example nova-api/resource/actions?relationshipType=&display=index&resources[]=12345 or ...&resources=all - we can just use $request->resources.

3. HOWEVER, for dependsOn fields this resources field isn't sent through at all with the PATCH nova-api/resource/action?action=action-name&editing=true&editMode=create&field=... call.

the very hacky solution i have implemented is to have a Hidden field 'action_resources' that contains the resources ids, and have the dependsOn field also depends on action_resources as well as whatever fields it really dependsOn. even though the dependsOn closure doesn't actually use this action_resources field, this crucially means that $request->action_resources will be included with the PATCH nova-api/resource/action?action=action-name&editing=true&editMode=create&field=... call.

```
Hidden::make('Action resources')->default($request->resources ?? $request->action_resources ?? $this->model($request)?->getKey()),
Select::make('Parent field')->options(['hide' => 'Hide', 'show' => 'Show'])->displayUsingLabels(),
Text::make('Child field')->dependsOn(['parent_field', 'action_resources'], function (Text $field, NovaRequest $request, FormData $formData) {
	if ($parent_field == 'show') {
		$field->show();
	} else {
		$field->hide();
	}
}),
```

currently, the function i'm using to retrieve the selected model (or null for all/multiple models - which can be changed as necessary) is:

```
protected function model(NovaRequest $request)
{
	if ($this->_model) {
		return $this->_model; // retrieve from stored model for per-resource actions in resource list
	}

	$resources = null;
	if ($request->resources) {
		$resources = $request->resources; // retrieve from request object for checkbox-selected resources
	} else if ($request->action_resources) {
		$resources = $request->action_resources; // retrieve from dependency for dependsOn field changes
		if (is_string($resources) && $resources != 'all') {
			$resources = json_decode($resources, true); // decode arrays from Hidden field which submits as a string
		}
	}

	// find real model or return null for multiple/'all' resources selected
	$modelClass = $request->model();
	if ($resources === null || $resources === 'all') {
		return null;
	} else if (is_array($resources)) {
		if (count($resources) == 1) {
			return $modelClass::find($resources[0]);
		} else {
			return null;
		}
	} else {
		return $modelClass::find($resources);
	}

	return null;
}
```

so, if the real resources parameter could be added to the PATCH nova-api/resource/action?action=action-name&editing=true&editMode=create&field=... call, it would do away with the hidden field, the extra dependsOn requirement, and the action_resources parsing in the model($request) call.

the only downside i can see is if you have a real field called 'resources' itself, but anyway when you actually confirm the action and it calls nova-api/resource/action?action=action-name&pivotAction=false&search=&filters=..., it will submit this field value as 'resources' anyway so it's not recommended to use this, and maybe even should throw an exception in Action when it finds a field called 'resources' perhaps?

anyway i hope this is useful? if not for whatever reason, i'll keep using my hack and of course anyone else who finds this is free to use the same setup for now :)

**Comment by puzzledmonkey** [comment]
i guess i forgot add to this - the whole point of my setup is the ability to use
```
public function fields(NovaRequest $request)
{
    $selected = $this->model($request);
    // ...
```
and then change the fields result based on this value :)

**Comment by crynobone** [comment, marked as answer]
Duplicate of https://github.com/laravel/nova-issues/issues/5567

**Comment by puzzledmonkey** [comment]
is this fixed in Nova 4.25.0? in the release notes it claims:

"Fixed an issue where selected resources were missing on action fields."

but i am still seeing the same issue with 4.25.1

**Comment by puzzledmonkey** [comment]
Closing due to apathy
